File: python/searching/search_records.py
#! /usr/bin/env python3

import json
import os
import sys
from pathlib import Path

from senzing import SzBadInputError, SzError, SzRetryableError, SzUnrecoverableError
from senzing_core import SzAbstractFactoryCore

INSTANCE_NAME = Path(__file__).stem
SETTINGS = os.getenv("SENZING_ENGINE_CONFIGURATION_JSON", "{}")

search_records = [
    {
        "NAME_FULL": "Susan Moony",
        "DATE_OF_BIRTH": "15/6/1998",
        "SSN_NUMBER": "521212123",
    },
    {
        "NAME_FIRST": "Robert",
        "NAME_LAST": "Smith",
        "ADDR_FULL": "123 Main Street Las Vegas NV 89132",
    },
    {
        "NAME_FIRST": "Makio",
        "NAME_LAST": "Yamanaka",
        "ADDR_FULL": "787 Rotary Drive Rotorville FL 78720",
    },
]


def mock_logger(level, error, error_record=None):
    print(f"\n{level}: {error.__class__.__name__} - {error}", file=sys.stderr)
    if error_record:
        print(f"{error_record}", file=sys.stderr)


def searcher(engine):
    for search_record in search_records:
        try:
            record_str = json.dumps(search_record)
            response = engine.search_by_attributes(record_str)
        except (SzBadInputError, json.JSONDecodeError) as err:
            mock_logger("ERROR", err, record_str)
        except SzRetryableError as err:
            mock_logger("WARN", err, record_str)
        except (SzUnrecoverableError, SzError) as err:
            mock_logger("CRITICAL", err, record_str)
            raise err

        print(f"\n------ Searched: {record_str}", flush=True)
        print(f"\n{response}", flush=True)


try:
    sz_factory = SzAbstractFactoryCore(INSTANCE_NAME, SETTINGS, verbose_logging=False)
    sz_engine = sz_factory.create_engine()
    searcher(sz_engine)
except SzError as err:
    mock_logger("CRITICAL", err)

File: java/snippets/searching/SearchRecords.java
package searching;

import java.io.StringReader;
import java.util.*;
import javax.json.*;

import com.senzing.sdk.*;
import com.senzing.sdk.core.SzCoreEnvironment;

import static com.senzing.sdk.SzFlag.*;

/**
 * Provides a simple example of searching for entities in the Senzing repository.
 */
public class SearchRecords {
    public static void main(String[] args) {
        // get the senzing repository settings
        String settings = System.getenv("SENZING_ENGINE_CONFIGURATION_JSON");
        if (settings == null) {
            System.err.println("Unable to get settings.");
            throw new IllegalArgumentException("Unable to get settings");
        }

        // create a descriptive instance name (can be anything)
        String instanceName = SearchRecords.class.getSimpleName();

        // initialize the Senzing environment
        SzEnvironment env = SzCoreEnvironment.newBuilder()
            .settings(settings)
            .instanceName(instanceName)
            .verboseLogging(false)
            .build();
        
        try {
            // get the engine from the environment
            SzEngine engine = env.getEngine();

            // loop through the example records and add them to the repository
            for (String criteria : getSearchCriteria()) {
                // call the searchByAttributes() function with default flags
                String result = engine.searchByAttributes(
                    criteria, SZ_SEARCH_BY_ATTRIBUTES_DEFAULT_FLAGS);
                
                JsonObject jsonObj = Json.createReader(
                    new StringReader(result)).readObject();

                System.out.println();
                JsonArray jsonArr = jsonObj.getJsonArray("RESOLVED_ENTITIES");
                if (jsonArr.size() == 0) {
                    System.out.println("No results for criteria: " + criteria);
                } else {
                    System.out.println("Results for criteria: " + criteria);
                    for (JsonObject obj : jsonArr.getValuesAs(JsonObject.class)) {
                        obj = obj.getJsonObject("ENTITY");
                        obj = obj.getJsonObject("RESOLVED_ENTITY");
                        long    entityId    = obj.getJsonNumber("ENTITY_ID").longValue();
                        String  name        = obj.getString("ENTITY_NAME", null);
                        System.out.println(entityId + ": " + name);
                    }
                }
                System.out.flush();
            }

        } catch (SzException e) {
            // handle any exception that may have occurred
            System.err.println("Senzing Error Message : " + e.getMessage());
            System.err.println("Senzing Error Code    : " + e.getErrorCode());
            e.printStackTrace();
            throw new RuntimeException(e);

        } catch (Exception e) {
            e.printStackTrace();
            if (e instanceof RuntimeException) {
                throw ((RuntimeException) e);
            }
            throw new RuntimeException(e);

        } finally {
            // IMPORTANT: make sure to destroy the environment
            env.destroy();
        }

    }

    /**
     * This is a support method for providing a list of criteria to search on.
     * 
     * @return A {@link List} {@link String} JSON text values desribing the
     *         sets of criteria with which to search.
     */
    public static List<String> getSearchCriteria() {
        List<String> records = new LinkedList<>();
        records.add(
            """
            {
                "NAME_FULL": "Susan Moony",
                "DATE_OF_BIRTH": "15/6/1998",
                "SSN_NUMBER": "521212123"
            }
            """);
        
        records.add(
            """
            {
                "NAME_FIRST": "Robert",
                "NAME_LAST": "Smith",
                "ADDR_FULL": "123 Main Street Las Vegas NV 89132"
            }
            """);
        
        records.add(
            """
            {
                "NAME_FIRST": "Makio",
                "NAME_LAST": "Yamanaka",
                "ADDR_FULL": "787 Rotary Drive Rotorville FL 78720"
            }
            """);

        return records;
    }
}